### ISSUE TITLE
Support OCI registries for provider and module distribution.

### ISSUE THREAD

**Issue by cube2222** [opentofu/opentofu#308]
Right now OpenTF relies on a specific registry implementation to host providers and modules. That works fine and is easily hostable using static files on something like S3, but generally, it's not trivial to self-host, and it's OpenTF+Terraform-specific.

There is however a common component used widely - OCI registries, like DockerHub or the GitHub Container Registry - that are basically perfect for the use-case, as they really are a generic storage for content-addressable-hash-identified blobs. If OpenTF supported these, then you could use any container registry you already have on hand to host providers and modules, either as a way to host custom ones, or in order to have an internal mirror of the public registry.

This does **not** mean that we'd somehow use docker to run providers, or anything like that. It's just using OCI registries to host the provider zipfiles and module tarballs. You can take a look at the [ORAS website](https://oras.land) to learn more about using OCI registries for generic artifact distribution.

I've already done a PoC of this for providers, the [diff of which you can find here](https://github.com/opentffoundation/opentf/compare/main...kubas-oci-images-experiment) and which also requires [some changes to the terraform-registry-address repository](https://github.com/cube2222/terraform-registry-address/compare/main...cube2222:terraform-registry-address:kubas-oci-images-experiment?expand=1).

Here's a short video of it:
![https://github.com/opentffoundation/opentf/assets/7013055/c9477fe9-df0c-416e-bb2e-c2572e6476da](https://github.com/opentffoundation/opentf/assets/7013055/c9477fe9-df0c-416e-bb2e-c2572e6476da)

I will prepare a proper RFC for this eventually, once I figure out a few details related to media types, artifact types, and the exact schema we should be using here.

**Comment by Magnitus-** [comment]
Are there well established OCI registries that you can easily operate though?

I ask, because we are on-prem and containers registry are one of the few things we are not operating ourselves (we still lean on Dockerhub for that).

I didn't find an s3-backed docker registry (based on the registry image) smooth sailing when I tried 2 years ago and Harbor is complex enough that the official instructions advocate using kubernetes to operate it (which you need to think really hard about if you want to use it as a foundational technology for your platform). I haven't given a good look at Dragonfly yet.

Ironically, compared to that, S3 hasn't been not bad actually with Minio or Ceph (the later come with most Openstack installations and the former is pretty nice to operate actually).

**Comment by ilmax** [comment]
@Magnitus- GitHub offers it as well. 
I'm not familiar with what you can use on prem, but you can always use your cloud supplier of choice and pay for a container registry.
This is different than using the terraform protocol because you can't easily run it on your own, there are some open source implementations out there but (at least when I last looked into it) no one of those were implementing the login protocol.
Also since the intention is not to replace the terraform protocol, but to offer an alternative, I don't see any downside.

**Comment by cube2222** [comment]
Yep, exactly what @ilmax said. There's many container registry implementations and it's a widely-used standard (on AWS you'd probably use the Elastic Container Registry, rather than S3). There's also an [official headless registry implementation by Docker](https://hub.docker.com/_/registry), useful for creating internal mirrors of the images you use. Though I believe [Harbor](https://goharbor.io) would be the fully-featured self-hosted option.

I think it makes much more sense to reuse all of this effort than have to author registries with a custom protocol, but as @ilmax said, this is meant to be an alternative, not a replacement for the registry protocol.

**Comment by Magnitus-** [comment]
> GitHub offers it as well.
> I'm not familiar with what you can use on prem, but you can always use your cloud supplier of choice and pay for a container registry.

We go with Dockerhub because it is cheap (25$/month) and expensing any recurring service is a pain here. However, the automation is non-existent (really old school dashboard experience which is ok if you have a small number of repositories in a small team).

Large cloud suppliers are less attractive on-prem, because they charge ridiculous egress fees (ie, 0.09$/GB for aws), although I should periodically take a look at what the more reasonable cloud suppliers (like Digital Ocean with its 0.01$/GB egress fee) are offering.

I'll have to look at Github.

But yes, at least there is choice.

> There's also an [official headless registry implementation by Docker](https://hub.docker.com/_/registry), useful for creating internal mirrors of the images you use, though I believe [Harbor](https://goharbor.io/) would be the fully-featured self-hosted option.

Yes, these are the two solutions I've looked at on-prem. The former was beautiful on my local machine using the filesystem for storage, but failed spectacularly once I tried to scale it against a Ceph object store. It looked so elegant and yet it didn't work and we were really short on time so we went with Dockerhub. It broke my heart.

Harbor looks very robust, though it also look like its trying hard to outdo Openstack, Kubernetes or Kafka in terms of operational complexity.

I'll have to try the registry implementation again, its been about 2 years now. Perhaps that has changed.

> as @ilmax said, this is meant to be an alternative, not a replacement for the registry protocol.

Well, if its optional, then more choice is always nice.

**Comment by ilmax** [comment]
@Magnitus- also worth nothing that a terraform modules shouldn't be as big as a regular container, most modules are a matter of KBs vs tens or hundred of MBs for regular container, so I am not expecting a cloud container registry used to store terraform modules to be very expensive

**Comment by Magnitus-** [comment]
> @Magnitus- also worth nothing that a terraform modules shouldn't be as big as a regular container, most modules are a matter of KBs vs tens or hundred of MBs for regular container, so I am not expecting a cloud container registry used to store terraform modules to be very expensive

I think its more like a couple of MBs, but fair point.

**Comment by stevehipwell** [comment]
> @Magnitus- also worth nothing that a terraform modules shouldn't be as big as a regular container, most modules are a matter of KBs vs tens or hundred of MBs for regular container, so I am not expecting a cloud container registry used to store terraform modules to be very expensive

Defining an ignore file like `.helmignore` for module publishing could make sure that the module only contains relevant files rather than the whole repo content.

**Comment by pataquets** [comment]
On the costs side, I'd like to bring to your attention a no-egress-fee object storage, in case it might be useful later:
[Cloudflare R2 | Zero Egress Distributed Object Storage](https://www.cloudflare.com/developer-platform/r2/)
Just my 2c.

(Disclaimer: I have no affiliation with CF, other than customer/user)

**Comment by itspngu** [comment]
Chiming in to express my support for this proposal, I've brought it up on hashicorp/terraform in the past, some possibly useful additional context and information can be found there: https://github.com/hashicorp/terraform/issues/31463

**Comment by cube2222** [comment]
@itspngu Glad to hear that!

If you can, please edit your comment to describe the issue here in place (or the important points from it), we generally try for the issue discussion here to be self-contained and not link to legacy Terraform issues / PRs / documents.

**Comment by itspngu** [comment]
> please edit your comment to describe the issue here in place (or the important points from it), we generally try for the issue discussion here to be self-contained 

I've already opened #376 (and closed it again after finding this issue), the contents are replicated there. However, the original issue's comments/discussion contain some useful additional information, and it'd feel weird to copy a series of posts by other people here if that makes any sense.

**Comment by pdecat** [comment]
Gitlab's container registry supports OCI pretty well, I'm using it for hosting one of my custom Steampipe plugin.

**Comment by Magnitus-** [comment]
> Defining an ignore file like `.helmignore` for module publishing could make sure that the module only contains relevant files rather than the whole repo content.

I should have re-read that. I was thinking about providers (which is what we tend to use registries for) to be in MBs. Modules, which are just hcl files, would indeed be measured more in KBs.

**Comment by buzzsurfr** [comment]
> Are there well established OCI registries that you can easily operate though?
> 
> I ask, because we are on-prem and containers registry are one of the few things we are not operating ourselves (we still lean on Dockerhub for that).

[Harbor](https://goharbor.io) is the registry we're using outside of cloud. It's a CNCF Graduated project and is pretty solid.

I'd contribute to this PR to help with the OCI bits and to get it working/test on Harbor.

**Comment by huix11** [comment]
👍👍👍👍

**Comment by cube2222** [comment]
We'd appreciate the community adding reactions to the post, or posting comments with their use-cases to support the implementation of this.

**Comment by ilmax** [comment]
@cube2222 do this need an RFC in order to be discussed and (hopefully) accepted and implemented?

**Comment by janosdebugs** [comment]
Hey @ilmax, just had a quick chat with @cube2222 , the intent here is that we'd like to hear a use case where the current registry is not sufficient. An RFC is not needed at this point, I believe.

**Comment by ilmax** [comment]
Hey @janosdebugs I can explain my use case if you want

**Comment by janosdebugs** [comment]
@ilmax please do!

**Comment by stevehipwell** [comment]
@janosdebugs OCI registries are a ubiquitous pattern with an effectively zero barrier to entry (GHCR). What alternative approach offers this ease of use? I think the Helm community experience could be used as a case in point, and in that scenario there was already a relatively simple self hosted option with automation support.

**Comment by janosdebugs** [comment]
Hey @stevehipwell, thank you. We are looking for specific things that not having OCI registries as distribution methods prevent OpenTofu users from doing.

**Comment by buzzsurfr** [comment]
> Hey @stevehipwell, thank you. We are looking for specific things that not having OCI registries as a distribution method prevent OpenTofu users from doing.

For me, it's less about not having OCI registries and more about having "yet another registry and format" that I have to maintain. I don't think the current mechanism needs to be deprecated, but make OCI an option. This relieves a team from having to provision, configure, and deploy another registry.

Helm is a good example here because like tofu, a helm registry was a static HTTP site with a specific document structure. When they moved to OCI (though it was less of a move and more like a target audience) then they didn't have to manage a separate registry.

GHCR is one of many [compatible OCI registries](https://oras.land/docs/compatible_oci_registries/), meaning hosting a new tofu package would be part of the pipeline to release a new package--the deployment of the package is done by OCI.

I (and others) pressed HC about Terraform using an OCI registry. They were never motivated because it competes with TF enterprise. Tofu doesn't have that issue.

**Comment by buzzsurfr** [comment]
If it's a question of contributors to make the changes, then I volunteer as tribute and encourage others to react to this comment as a sign of those who can work on it.

**Comment by janosdebugs** [comment]
Hey @buzzsurfr thank you for offering to implement this, it is much appreciated. When the core team looks at issues, there are a few points that go into that consideration:

1. Is it possible to implement this on a technical level? (The answer for OCI is yes.)
2. Does the feature cause technical debt? (No, this likely wouldn't.)
3. Is there someone who would do the work. (Yes, thank you very much for the offer!)
4. Is there enough capacity on the core team to support a contributor in developing the feature and also do however many rounds of reviews are needed. (We are a bit light on capacity, but if there's enough need, we'll make it happen.)
5. Is there enough community need for the feature to justify the continued maintenance of the feature? (This is what we want to find out.)

So, in essence, it is a question if this feature enables someone to do something they weren't able to do before and that it does so for enough people to justify adding the necessary dependencies and maintaining (testing, upgrading, bugfixing) a likely significant amount of code.

Personally, I would love to support OCI (and also providers as containers, see #1138), but given the over 300,000 lines of Go code in this project, adding new things is something of a balancing act.

**Comment by ilmax** [comment]
As of today, (unless I'm missing something obvious) beside Terraform Cloud or [GitLab Terraform Module Registry](https://docs.gitlab.com/ee/user/packages/terraform_module_registry/), there's no good solution to share private modules beside git references. Git references comes with some downside in GitHub.

I see OCI as an answer in this area because:

- It's a standard specification
- There are plenty of offerings for OCI compatible registries
- It will simplify private module sharing
- If you work with containers, you most likely already have one or more
- It's adding yet another option to module source

I understand that adding new feature is a balancing act, but IMO this will be a killer feature and, like @buzzsurfr said above, HashiCorp was not interested into building this one due to it's competing product offering.

Give that OpenTofu has been submitted to the CNCF sandbox, OCI support seems like a great fit 😄

**Comment by ilmax** [comment]
@Yantrio the first 2 do not implement the [Login Protocol](https://developer.hashicorp.com/terraform/internals/login-protocol), _terrareg_ seems like a good alternative but just looking at the user guide is way more complicated than creating a container registry OCI compliant.

Also no one was taking about switching to OCI, time will tell if OCI is a better solution than other ways of using module source. 
In my opinion OCI will be suited for sharing private module and eventually just to add redundancy on public module for paranoid, that's the use case I can imagine so I don't see it replacing anything else.

**Comment by DavidGamba** [comment]
Our CI environment is constantly hitting Github API rate limits so having a way to deploy to the OCI registry might alleviate that.
IMHO the same should be done for the provider registry. Proxying through a private OCI registry so that images are auto cached and you only go outside for new ones.

I started using the [Oras project](https://oras.land/) to put assets in my container registry (ECR) (for example golang binaries) and you could use it to vendor your terraform modules that way but something built-in would be way better.

**Comment by janosdebugs** [comment]
Hey @DavidGamba thank you, that is very good info. Could you paste the error message for that? I want to see if we can do something short-term about it.

**Comment by damnedOperator** [comment]
We're starting the migration from bicep to tofu and one of the major downsides in the discussion that came up was the lack of OCI module hosting support that is present in bicep, but neither in TF itself or tofu. We're in year 2 of the project right now and changing the way we store our IaC artifacts to a completely new type of registry is finnicky at last. 
We want to leverage Azure Container Registry for hosting those OCI artifacts, so we can also use our Workload Identity enabled build agents to handle the whole authentication process, which is also a problem when working with Git hosted Modules in Azure DevOps Services.
So... +1 from my side for supporting OCI artifacts

**Comment by buzzsurfr** [comment]
1 and 2 are probably up for discussion as there’s a few of us with ideas.
3. OCI includes the ability to track commit IDs, so that can be built in.
4. OCI registries have a known API, so we could either use that or write a plugin/class to transform that layer.

I’m down to contribute on the RFC/code portion as well (as I suspect many are). Going on vacation in a few but I’ll check back after and if no one else starts it, I can take a first pass.

**Comment by cam72cam** [comment]
Unasigning @cube2222 so someone else can write a RFC based on the POC

**Comment by eunanhardy** [comment]
I am currently working on a POC around this issue to support oci based images in my own module registry [terrapak](https://github.com/eunanhardy/terrapak), I am happy to contribute my findings to an RFC. Some thoughts I have based on raised questions. 

1. including `oci://<reigstry-url>/name` be supported within module source with version being used to specify tag
2. supporting inclusion of local module dependencies, e.g. `../../my-module` being bundled when the target module is packaged
3. If you want to store a commit id, the oci spec supports this through the `org.opencontainers.image.revision` annotation within the image manifest. it could be included if a flag is present for automation. 
4. Maybe consider following current conventions around the API 
```bash
tofu registry login | pull | push | add | rm
```

**Comment by damnedOperator** [comment]
> I think the idea would be to support OCI registries without an oci:// prefix purely based on the /.well-known URL prefix and detect what kind of a registry it is.

In my opinion, an `oci://` prefix would help with readability. We have "protocol identifiers" for everything else, so why not for OCI registries?

**Comment by eunanhardy** [comment]
> Hey @eunanhardy thank you very much. 
> 
> 
> 
> 1. I think the idea would be to support OCI registries without an `oci://` prefix purely based on the `/.well-known` URL prefix and detect what kind of a registry it is.
> 
> 2. _
> 
> 3. The problem is that the label can be manipulated and a git SHA can't easily. This gives the user the certainty that no supply chain attack has taken place.
> 
> 
> 
> 

1. Similar to what @damnedOperator said protocol identifies are already in use and developing a new module getter with the prefix pattern would be easier. What is the primary reason or benefit  for wanting a new behaviour for OCI.
3. The problem I have with this is that it assumes the module being packaged exists within a git repo which although likely should not be assumed for default behaviour. Manifest annotations also have better support for surfacing this information in third party registries.

**Comment by cube2222** [comment]
Regarding the RFC, one of the bigger unknowns is the actual structure/schema of the OCI artifact.

OCI storage mostly consists of content-addressable blobs, manifests that are jsons and often reference blobs or other manifests, as well as mutable tags to reference blobs or manifests. E.g. multi-platform docker images have a central manifest (for the overall image) that points to one manifest per platform (among other metadata). Each per-platform manifest contains a list of references to its layers (among other metadata). Layers are blobs which contain tarballs of directory structures.

In my PoC I made a somewhat sensible structure that supports providers for many operating systems and architectures, but it wasn't well thought-out. It was loosely inspired by multi-platform docker images, though simplified.

It would be good if someone (either as a comment, or as part of an RFC) would suggest a concrete schema and structure for the manifests we'd store. What should the mediaType and artifactType be. Where should we reuse existing ones, and where should we introduce our own.

Maybe actually just using the format used by container images is the way to go (after all, a provider is in fact a tarball of files with an entrypoint, which is **almost** what a container image is if you squint hard enough)? Or maybe not.

**Comment by programmer04** [comment]
It may be an interesting read for inspiration - [how OCI images are used to ship WASM plugin to e.g. Istio](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/how-to-build-oci-images.md) - it's basically a `FROM SCRATCH` image with a single layer that contains binary

**Comment by eunanhardy** [comment]
Submitted a RFC for this issue [#1672](https://github.com/opentofu/opentofu/issues/1672)

**Comment by DavidGamba** [comment]
The CUE project just landed their implementation of modules. They have some nicely written documentation around it:
https://github.com/cue-lang/proposal/blob/main/designs/2330-modules-v2.md

They open sourced their OCI code here: https://github.com/cue-labs/oci

**Comment by cube2222** [comment]
> The only concern is that someone might misunderstand that layout as something where they can ship additional files alongside the binary, which is, of course, not the case.

Is it not the case? You can already do that by including additional files in the provider tarballs, and I think that's fine, and nothing prohibits you from using them, and the provider using those files when running?

Either way, I really like that `FROM SCRATCH` approach. It's similar to how AWS Lambda layers work too.

**Comment by eunanhardy** [comment]
I like the `FROM SCRATCH` approach but I think that the complexity with packing and unpacking images is about the same as implementing the image spec natively within Tofu. It will also provide the benefit of owning the entire lifecycle allowing us to build on top of this feature in the future.

**Comment by nick-williams-spirit** [comment]
I might be misunderstanding the conversation about structure/schema of the OCI artifact. But is this the sort of situation ORAS is trying to standardize?

https://oras.land/docs/

Helm uses this under the hood for storing helm charts in OCI registries. It feels like a similar concept?

**Comment by siegenthalerroger** [comment]
Seconding @nick-williams-spirit on the ORAS "standardisation" that uses media-type. 

An alternative project that I haven't seen mentioned is FluxCD's usage of OCI to store kustomize bundles. They're looking to push this to kustomize upstream so their layer design might be worth looking at. At the end of the day from what I understand it's quite ORAS compatible in the sense that it's a tarball with the correct labelling of the layer.

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

We've been working on this stuff over in https://github.com/opentofu/opentofu/issues/2540, and at this point all of the code is in `main` and there's only release preparation tasks and other-codebase-related tasks left on the list, so I'm going to close this issue now.

If you're interested we'd love help with testing this using the v1.10.0-alpha2 release, or with the subsequent beta that should be out pretty soon once some other work is finished. There's some information about this in the earlier alpha release blog posts:

- [Provider distribution through OCI Registries](https://opentofu.org/blog/help-us-test-opentofu-1-10-0-alpha1/#provider-distribution-through-oci-registries)
- [OCI Registry integration](https://opentofu.org/blog/help-us-test-opentofu-1-10-0-alpha2/#oci-registry-integration)
- [Draft documentation](https://opentofu.org/docs/main/cli/oci_registries/) (after final release this will move to [OCI Registry Integrations](https://opentofu.org/docs/cli/oci_registries/), once the v1.10.0 docs become "current" on the website.

If you try it and have specific feedback you'd like to share, please **open a new issue** rather than comment on this one, because it'll be very hard to track multiple different threads of feedback all in this one issue. Thanks!
